neqo-transport/src/connection.rs

```diff
@@ -1542,16 +1629,44 @@ impl Connection {
         }
     }
 
-    fn validate_odcid(&mut self) -> Res<()> {
-        // Here we drop our Retry state then validate it.
-        if let Some(info) = self.retry_info.take() {
+    /// Process the final set of transport parameters.
+    fn process_tps(&mut self) -> Res<()> {
+        self.validate_cids()?;
+        if let Some(token) = self
+            .tps
+            .borrow()
+            .remote
+            .as_ref()
+            .unwrap()
+            .get_bytes(tparams::STATELESS_RESET_TOKEN)
+        {
+            let reset_token = <[u8; 16]>::try_from(token).unwrap().to_owned();
+            self.path.as_mut().unwrap().set_reset_token(reset_token);
+        }
+        self.set_initial_limits();
+        Ok(())
+    }
+
+    fn validate_cids(&mut self) -> Res<()> {
+        match self.quic_version {
+            QuicVersion::Draft27 => self.validate_cids_draft_27(),
+            _ => self.validate_cids_draft_28_plus(),
+        }
+    }
+
+    fn validate_cids_draft_27(&mut self) -> Res<()> {
+        if let Some(info) = &self.retry_info {
             if info.token.is_empty() {
                 Ok(())
             } else {
                 let tph = self.tps.borrow();
-                let tp = tph.remote().get_bytes(tparams::ORIGINAL_CONNECTION_ID);
+                let tp = tph
+                    .remote()
+                    .get_bytes(tparams::ORIGINAL_DESTINATION_CONNECTION_ID);
                 if let Some(odcid_tp) = tp {
-                    if odcid_tp[..] == info.odcid[..] {
```

Comment:
This test can probably be easier.  `get_bytes` returns `Option<&[u8]>`, which we can compare directly with `self.original_destination_cid.as_ref().map(ConnectionId::borrow)`.  That removes a whole layer of if statements.

```rust
if self.original_destination_cid.as_ref().map(ConnectionId::borrow) == tp {
  Ok(())
} else {
  Err(Error::InvalidRetry)
}
```

The alternative is to convert to `ConnectionIdRef`, but it's less clear:

```rust
if self.original_destination_cid.as_ref().map(ConnectionId::as_cid_ref) == tp.map(ConnectionIdRef::from) {
// ...
```

Reply:
The first seems to need `std::borrow::Borrow`, and that gets ugly because we're already using another trait which defines a `borrow()` so 20 locations start needing disambiguation. So I went with your second option, which seems to work nicely and gets rid of the if layer.